### PR TITLE
Update the owning-inventory for existing objects during migration

### LINKED ISSUE #1306: Regression: pruning fails for ResourceGroup inventory object
* After a recent PR: https://github.com/kubernetes-sigs/cli-utils/pull/291, pruning fails for `ResourceGroup` inventory object after migration.

Steps to reproduce:
1. set environment variable `export RESOURCE_GROUP_INVENTORY=1`
2. start/restart cluster with kind: `kind delete cluster; kind create cluster`
3. setup the "app": `cp -f e2e/live/testdata/template-rg-namespace.yaml e2e/live/testdata/migrate-case-1a/inventory-template.yaml`
5. apply the "app" with ConfigMap inventory: `kpt live apply e2e/live/testdata/migrate-case-1a`
6. migrate the "app" to ResouceGroup: `kpt live migrate e2e/live/testdata/migrate-case-1a`
6. copy Kptfile to next version: `cp -f e2e/live/testdata/migrate-case-1a/Kptfile e2e/live/testdata/migrate-case-1b`
7. apply/prune the next version of the "app":  `kpt live apply e2e/live/testdata/migrate-case-1b`

Output:
```
namespace/test-rg-namespace unchanged
pod/pod-b unchanged
pod/pod-c unchanged
pod/pod-d created
4 resource(s) applied. 1 created, 3 unchanged, 0 configured, 0 failed
pod/pod-c prune skipped
pod/pod-b prune skipped
pod/pod-a prune skipped
namespace/test-rg-namespace prune skipped
0 resource(s) pruned, 4 skipped, 0 failed
```

All the resources are erroneously `prune skipped`

Expected Output:
```
namespace/test-rg-namespace unchanged
pod/pod-b unchanged
pod/pod-c unchanged
pod/pod-d created
4 resource(s) applied. 1 created, 3 unchanged, 0 configured, 0 failed
pod/pod-a pruned
1 resource(s) pruned, 0 skipped, 0 failed
```

Reason: It looks like a new `CanPrune` function is erroneously identifying all resources as having a different inventory object
```
I1215 16:24:38.369507 1266578 prune.go:230] skip pruning object that doesn't belong to current inventory: 5b339645-7b28-4de2-a1f9-5f5ce0763140
pod/pod-c prune skipped
I1215 16:24:38.371694 1266578 prune.go:230] skip pruning object that doesn't belong to current inventory: 1c16ea30-5583-499a-a0ea-1283e4292f13
pod/pod-b prune skipped
I1215 16:24:38.373388 1266578 prune.go:230] skip pruning object that doesn't belong to current inventory: 506cc2ed-6aba-4d7c-a55c-966a21d4b8cf
pod/pod-a prune skipped
I1215 16:24:38.374488 1266578 prune.go:230] skip pruning object that doesn't belong to current inventory: 02c9183e-7903-4b35-9af1-986b5eb0154a

```

### PR DESCRIPTION
**Pull request by Liujingfang1**
After we add the InventoryPolicy in the cli-utils library, we need to update the migrate command so that the owning-inventory annotations are pointed to the ResourceGroup object. The change is to loop over the resources that are included in the ConfigMap inventory object and change their `owning-inventory` annotation to the new ResourceGroup CR.

When updating the `owning-inventory` annotation, we also update the last-applied-configuration annotation.

Fix #1306 This is a different approach compared with #1307.

### CODE REVIEW

File: commands/migratecmd.go
Comment (mortent) [inline commands/migratecmd.go:270]: So in this case we add the `owning-inventory` annotation regardless of whether the resource had one already right?
If something should fail during migration and kpt crashes just before this line, would re-running the migration still result in the correct end state? I think so, but I just want to make sure we have thought that through.

File: pkg/client/client.go
Comment (mortent) [inline pkg/client/client.go:42]: Does this result in an `UPDATE` or a `PATCH` call to the apiserver? Just want to make sure we do this the same way as a regular apply operation.

**Review by mortent — COMMENTED** [review]
Some small questions, but I think this looks good. I think this approach is simpler and safer than in #1307 
I'll let @seans3 take a look too

File: commands/migratecmd.go
Comment (Liujingfang1) [reply commands/migratecmd.go:270]: This function checks if the resource has the `owning-inventory` annotation. If it doesn't have or if the `owning-inventory` points to the previous ConfigMap inventory object, it updates the `owning-inventory` to the new ResourceGroup object. The reason of doing this is that the `owning-inventory` might be other inventory object and we can't always change it to the ResourceGroup object during migration. If the `owning-inventory` is already the ResourceGroup object, we don't need to update it; if the `owning-inventory` is another object, the migration can't adopt it to the current ResourceGroup object. 

Since the current logic is to create the ResourceGroup CR first, then update the `owning-inventory` annotation. In the case you mentioned, if the ResourceGroup object is created, re-running the migration would skip the part for adding `owning-inventory` annotation.  If the ResourceGroup object wasn't created, re-running the migration would result in the correct end state.

If we change it to first update the `owning-inventory` annotations; then create the ResourceGroup CR. If there is any failure during the migration, re-running the migration would fix it.

File: pkg/client/client.go
Comment (Liujingfang1) [reply pkg/client/client.go:42]: It's an `UPDATE` call. I didn't use the `PATCH` call since we should only change the `metadata.annotation`.  Only two annotations are updated: the `owning-inventory` and the last-applied-config annotation. If we compute them correctly, we don't need to use the `PATCH` call.

File: commands/migratecmd.go
Comment (Liujingfang1) [reply commands/migratecmd.go]: Changed the order of the two operations. PTAL.

File: pkg/client/client.go
Comment (seans3) [reply pkg/client/client.go:42]: For more context: the `Update` call translates to a REST `Put` with the entire object in the body.

**Review by seans3 — APPROVED** [review]
/lgtm

**Comment by seans3** [comment]
Should we squash the commits before merging?

**Comment by Liujingfang1** [comment]
@seans3 I squashed the commits. Let's wait for the tests to pass.

**Pull request merged**
